substrate/frame/assets/src/types.rs

```diff
@@ -35,6 +35,31 @@ pub(super) type AssetAccountOf<T, I> = AssetAccount<
 pub(super) type ExistenceReasonOf<T, I> =
 	ExistenceReason<DepositBalanceOf<T, I>, <T as SystemConfig>::AccountId>;
 
+/// Whether the set metadata operation is coming from a `force_origin` or not.
+// If it's a force origin then `is_frozen` is provided for the asset as well.
+pub(super) enum MetadataRequestOf<AccountId> {
+	ForceOrigin(bool),
+	Origin(AccountId),
+}
```

Comment:
IMO this is tries to conflate too many things and doesn't express the information clearly (and the below `impl` likewise returns types that aren't intuitive to the user). I think it would be a better API to remove this type and change the signature of `do_set_metadata` to something like this, and leave it up to the caller to supply the correct arguments when calling it:
```rust
pub(super) fn do_set_metadata(
	id: T::AssetId,
	// `Some` if called by an origin that converts to an `AccountId`.
	// Must equal the asset class's `owner`. `None` if called by a
	// privileged function that bypasses any permissions (that is,
	// the origin has already been checked and has the privilege to
	// call this function).
	maybe_owner: Option<AccountId>,
	name: Vec<u8>,
	symbol: Vec<u8>,
	decimals: u8,
	// Whether or not the asset class is frozen.
	is_frozen: bool,
) -> DispatchResult 
```